### DISCUSSION TITLE
Use custom onChange for triggering validation in a different field

### DISCUSSION THREAD

**Discussion by LauSam09** [react-hook-form/react-hook-form#8865] [Q&A] [answered]
# The problem
My use-case is wanting to validate a field based on the value of another field. E.g. a user must enter a `first name` if a `last name` is entered, but both can be left blank and be valid.

# Minimal example
In this sandbox, I've fixed the issue by using the API as suggested in [this feature request](https://github.com/react-hook-form/react-hook-form/issues/4751) and implemented last year. The problematic code is commented out.
https://codesandbox.io/s/conditionally-required-giv3k5

# Explanation
I already knew that I could use `getValues` to retrieve the value for the `lastName` field in the `firstName` validation code, and I knew that I needed to trigger the `firstName` validation from the `lastName` field to get the validation `onChange` behaviour. What I didn't realise, was that `register` has its own `onChange` implementation, and I was inadvertently overriding this by adding `onChange={() => trigger("firstName")` directly to my `lastName` input. This results in stale values causing undesirable behaviour, and potentially other issues as well, e.g. see the GIF below. Someone else appears to have made a similar mistake in the issue linked above.

![conditionally_required](https://user-images.githubusercontent.com/24778857/184902642-f77a0928-67b0-47c1-9bed-fa505dfc090a.gif)

Obviously overriding this prop is not intended, and we should instead supply our custom `onChange` function as an argument for `register`. However, aside from a typescript error if you're lucky enough to put your `onChange` _above_ the spread `register` call (I wasn't), it's tough to see the problem! Example error if you're luckier than me:
`'onChange' is specified more than once, so this usage will be overwritten. ts(2783`

So I've solved this issue myself, but it took me a while to figure it out and I don't think this is documented anywhere so wanted to save others the trouble.

**Comment by LauSam09** [comment]
As written above, the solution is to supply our custom `onChange` implementation that triggers field validation to the `register` function, instead of supplying it directly ourselves:

```diff
<input
+  {...register("lastName", { onChange: () => trigger("firstName") })}
-  {...register("lastName")}
-  onChange={() => trigger("firstName")}
/>
```

**Comment by Moshyfawn** [comment, marked as answer]
Check out [the `deps` API](https://react-hook-form.com/api/useform/register#options) as well

**Reply by LauSam09** [reply]
Thanks! I've gone with this approach and I'll mark this as the answer instead. Hopefully if there are more niche situations that can't be handled using `deps` then my answer will still be useful for people.

**Reply by Moshyfawn** [reply]
The `deps` API is a convenience wrapper around the `trigger` method calls, so you correct: for most of use-cases you can simply define the dependent validation fields, but if you have a more "custom" behavior, reach to manually call the `trigger` method whenever instead
